### ISSUE TITLE
Encoder embeddings: what carried, and what did not

### ISSUE THREAD

**Issue by znat** [znat/parrotflow#94]
## Summary

The plan leaves open whether a trained speech encoder (Parakeet's frames, mean-pooled and compared by cosine) beats the simple MFCC + DTW distance already in use for matching a spoken span against the voice bank. We measured it on 2026-08-10 on three tests: the encoder wins clearly on two, but on the test that matters most — scoring the actual proposals the vocabulary pass makes — it does not beat MFCC (AUC 0.825 vs. 0.810, which the per-term breakdown shows is a tie, not a win).

**The negative result is the finding, not a footnote.** The encoder should not replace MFCC as the shipped representation. It is still useful as a diagnostic instrument, because it is far more stable across microphones and recording sessions than MFCC — that stability is what the two positive tests show.

## Acceptance criteria

- [ ] MFCC + DTW stays the shipped representation. No change swaps in the encoder on the strength of this measurement alone.
- [ ] The encoder is not treated as validated for scoring decisions until the proposals test (arm 3 below) is re-run on a set covering all eleven terms — five of eleven had no correct proposal at all in this run and contributed nothing to the 0.810 / 0.825 numbers.
- [ ] The cross-microphone result (arm 1) is treated as a demonstration, not a rate, until at least ten cross-device pairs are measured against real per-device banks (tracked in #97).
- [ ] Every AUC already published in #74, #77, #82, #87 is annotated as a per-clip hold-out number, with the leave-one-session-out gap stated alongside it (measured: MFCC pooled 0.943 → 0.800, encoder pooled 0.982 → 0.950).

<details>
<summary>Full measurement detail: how the encoder was run, all three arms, and the size-matched control</summary>

### How the encoder is obtained — no app, no download

The encoder is already on the machine:

```
~/Library/Application Support/FluidAudio/Models/parakeet-ctc-110m-coreml/
    MelSpectrogram.mlmodelc
    AudioEncoder.mlmodelc
    vocab.json
```

Loaded straight with `coremltools`:

```python
import coremltools as ct
MODELS = Path.home() / "Library/Application Support/FluidAudio/Models/parakeet-ctc-110m-coreml"
mel_model = ct.models.CompiledMLModel(str(MODELS / "MelSpectrogram.mlmodelc"))
enc_model = ct.models.CompiledMLModel(str(MODELS / "AudioEncoder.mlmodelc"))

def encode(samples):
    buf = np.zeros(240000, dtype=np.float16)          # the export's fixed 15s window
    buf[:min(samples.size, 240000)] = samples[:240000].astype(np.float16)
    mel = mel_model.predict({"audio": buf})["melspectrogram_features"]
    return enc_model.predict({
        "melspectrogram_features": mel.astype(np.float16),
        "input_1": np.array([[[[0.0]]]], dtype=np.float16)})
```

`encoder_output_embeds` is the hidden state, `ctc_head_output` is the CTC logits. The pipeline was sanity-checked by greedy-decoding the logits back to a transcript, so the embeddings are known to be the real states and not garbage.

- The window is 15s fixed, 240,000 samples at 16 kHz.
- The window gives 188 encoder frames, so one frame is 79.8ms.
- No app run, no build, no download. Read-only against FluidAudio's cache.

### The constraint that decides how to use it — an isolated cut embeds to nothing

A half-second cut, zero-padded out to the 15s window, greedy-decodes to the empty string. Every such cut. Those embeddings were reading the padding, not the word.

The encoder is a conformer. It attends over the whole window, and 14.5s of silence is nothing it ever saw in training. Two runs of this experiment were thrown away before this was noticed.

So the encoder must run over the whole utterance, and the frames covering the span sliced out of it. `voice/observations.jsonl` already names the source clip and the span for every recording, which is what makes that possible. Clips longer than 15s get a 15s window centred on the span.

The MFCC baseline does not have this problem, which is part of why it survived the mistake.

A raw cosine is nearly useless. Mean-pooled encoder frames share a large common direction — every raw cosine in the first run sat between 0.78 and 0.90. Subtract the mean over the whole reference bank first. That is the same job CMVN does for the MFCC. All the numbers below are the centred cosine.

### Arm 1 — cross-microphone. The encoder un-inverted a pair MFCC got backwards

Two queries recorded on the MacBook microphone, scored against all eleven term banks, which are all RØDE:

- `08-39-07`, 8.88–9.36s — a genuine utterance of `Matthieu`.
- `08-39-24`, 9.12–9.68s — the ordinary word "matching", which the pass writes `Matthieu` over.

Distance to the `Matthieu` bank, and where `Matthieu` ranks among the eleven terms:

| | MFCC + DTW | rank | centred cosine | rank |
|---|---|---|---|---|
| true `Matthieu` | 3.452 | 3 / 11 | 0.3476 | 2 / 11 |
| false "matching" | **3.339** | **1 / 11** | 0.1895 | 3 / 11 |

MFCC has the pair backwards. The wrong word is nearer the bank than the right one, and MFCC ranks `Matthieu` first for the word that is not `Matthieu` and third for the word that is.

The encoder separates them the right way round, at every slice tried. The slice sweep, shifting the frame window from 2 frames early to 3 frames late and applying the same shift to the bank and to both queries:

| shift | cos, true | cos, false | rank true | rank false | DTW true | DTW false |
|---|---|---|---|---|---|---|
| −2 | 0.3414 | 0.1822 | 1 | 3 | 3.452 | 3.339 |
| −1 | 0.3395 | 0.1796 | 2 | 4 | 3.452 | 3.339 |
| +0 | 0.3476 | 0.1895 | 2 | 3 | 3.452 | 3.339 |
| +1 | 0.3031 | 0.2170 | 2 | 2 | 3.452 | 3.339 |
| +2 | 0.3518 | 0.2248 | 1 | 2 | 3.452 | 3.339 |
| +3 | 0.3614 | 0.2149 | 1 | 3 | 3.452 | 3.339 |

The true query wins at all six slices. The verdict does not depend on how the span is sliced into frames, which was the obvious way this could have been an artefact.

This is two clips. It is a demonstration, not a rate. What it demonstrates is that the failure mode exists and that the encoder does not have it here.

The same table is the evidence for per-microphone banks: #97.

### Arm 2 — leave-one-session-out over all 122 recordings

The published per-clip hold-out drops only the recordings cut from the query's own clip. A recording made minutes later, same session, same microphone, same room, is still free to vouch for it. No published number separates "same word" from "same session".

Leave-one-session-out holds out a whole day. For each term and each day it has: bank = the term's recordings from every other day; positives = the term's recordings from that day; negatives = every other term's recordings, against the same reduced bank. Pooled over the days. AUC as `scripts/reference-matching.py` computes it, ties half.

#### Encoder, centred cosine

| term | days | n | per-clip | per-session | gap | published |
|---|---|---|---|---|---|---|
| Arexvy | 2 | 7 | 1.000 | 0.974 | −0.026 | 0.936 |
| Claude | 1 | 6 | 0.986 | — | — | 0.844 |
| Matthieu | 3 | 11 | 0.971 | 0.920 | −0.051 | 0.848 |
| Mirza | 3 | 15 | 0.997 | 0.989 | −0.009 | 0.940 |
| Ollama | 2 | 7 | 0.986 | 0.986 | −0.000 | 0.874 |
| Praisy | 3 | 26 | 0.971 | 0.895 | −0.076 | — |
| Redcrawl | 1 | 8 | 0.985 | — | — | 0.966 |
| Redrock | 1 | 7 | 0.976 | — | — | 0.993 |
| Supabase | 3 | 11 | 0.971 | 0.913 | −0.058 | 1.000 |
| Tasmeen | 3 | 8 | 1.000 | 0.999 | −0.001 | 0.985 |
| Vercel | 4 | 16 | 0.985 | 0.984 | −0.001 | — |
| **pooled** | | | **0.982** | **0.950** | **−0.033** | 0.935 |

#### MFCC + DTW

| term | days | n | per-clip | per-session | gap | published |
|---|---|---|---|---|---|---|
| Arexvy | 2 | 7 | 0.957 | 0.891 | −0.066 | 0.936 |
| Claude | 1 | 6 | 0.885 | — | — | 0.844 |
| Matthieu | 3 | 11 | 0.883 | 0.832 | −0.051 | 0.848 |
| Mirza | 3 | 15 | 0.971 | 0.973 | +0.001 | 0.940 |
| Ollama | 2 | 7 | 0.891 | **0.502** | −0.389 | 0.874 |
| Praisy | 3 | 26 | 0.931 | **0.623** | −0.307 | — |
| Redcrawl | 1 | 8 | 0.984 | — | — | 0.966 |
| Redrock | 1 | 7 | 0.997 | — | — | 0.993 |
| Supabase | 3 | 11 | 1.000 | 0.924 | −0.076 | 1.000 |
| Tasmeen | 3 | 8 | 0.951 | 0.816 | −0.135 | 0.985 |
| Vercel | 4 | 16 | 0.976 | 0.959 | −0.019 | — |
| **pooled** | | | **0.943** | **0.800** | **−0.143** | 0.935 |

Per-clip pools all 11 terms. Per-session pools the 8 terms with two or more days. `Ollama` at 0.502 under leave-one-session-out is chance.

MFCC loses 0.143 when the session is taken away. The encoder loses 0.033.

### The size-matched control — how much of that drop is the session?

Holding out a whole day does two things at once. It removes the same-session siblings, which is the point. It also shrinks the bank: `Praisy` keeps 24 of its 26 recordings on one day, so holding that day out leaves a bank of 2 against the 25 a per-clip hold-out leaves.

The control separates them. For every fold — positives from day D, bank of k recordings from the other days — the same positives are re-scored against a bank of the same size k drawn at random from all the term's other recordings, same-day siblings allowed, the query itself never in it. 30 draws, averaged, seed 7.

```
size-matched − LOSO   = the session
per-clip − size-matched = the bank size
```

| | session | bank size |
|---|---|---|
| encoder | **−0.023** | **+0.050** |
| MFCC + DTW | **+0.037** | **+0.091** |

Per term, MFCC:

| term | n | bank k | per-clip | size-matched | LOSO | session | size |
|---|---|---|---|---|---|---|---|
| Arexvy | 7 | 1–6 | 0.952 | 0.879 | 0.891 | −0.012 | +0.073 |
| Matthieu | 11 | 4–10 | 0.878 | 0.905 | 0.832 | +0.073 | −0.027 |
| Mirza | 15 | 7–13 | 0.971 | 0.972 | 0.973 | −0.001 | −0.001 |
| Ollama | 7 | 1–6 | 0.882 | 0.646 | 0.502 | +0.144 | +0.236 |
| Praisy | 26 | 2–25 | 0.929 | 0.677 | 0.623 | +0.054 | +0.251 |
| Supabase | 11 | 2–10 | 1.000 | 0.911 | 0.924 | −0.013 | +0.089 |
| Tasmeen | 8 | 2–7 | 0.950 | 0.874 | 0.816 | +0.058 | +0.075 |
| Vercel | 16 | 5–15 | 0.976 | 0.948 | 0.959 | −0.010 | +0.028 |

Per term, encoder:

| term | n | bank k | per-clip | size-matched | LOSO | session | size |
|---|---|---|---|---|---|---|---|
| Arexvy | 7 | 1–6 | 1.000 | 0.965 | 0.974 | −0.009 | +0.035 |
| Matthieu | 11 | 4–10 | 0.968 | 0.960 | 0.920 | +0.040 | +0.008 |
| Mirza | 15 | 7–13 | 0.997 | 0.988 | 0.989 | −0.001 | +0.009 |
| Ollama | 7 | 1–6 | 0.984 | 0.873 | 0.986 | −0.113 | +0.111 |
| Praisy | 26 | 2–25 | 0.971 | 0.844 | 0.895 | −0.051 | +0.127 |
| Supabase | 11 | 2–10 | 0.969 | 0.888 | 0.913 | −0.025 | +0.081 |
| Tasmeen | 8 | 2–7 | 1.000 | 0.980 | 0.999 | −0.020 | +0.020 |
| Vercel | 16 | 5–15 | 0.984 | 0.976 | 0.984 | −0.008 | +0.008 |

Read this carefully. Most of both drops is the bank getting smaller, not the session. For the encoder the session term is negative — a bank without the query's own session is, on average, no worse. For MFCC the session is worth +0.037, which is real but is a quarter of the raw −0.143.

This is also an independent argument for the abstain floor (#99): the bank size term is +0.05 to +0.09 AUC, which is larger than most margins this work argues about.

### Arm 3 — the one that did not carry

On the acoustic pass's own proposals, the advantage vanishes.

171 proposals over the 141-clip set that the label settles: 51 where the term really was said, 120 where it was not.

| statistic | AUC |
|---|---|
| MFCC, `distance / spread` | 0.810 |
| encoder, `(1 − cos) / (1 − cloud)` | **0.825** |

Per term, where both classes are present:

| term | said | not | MFCC | encoder |
|---|---|---|---|---|
| matthieu | 5 | 10 | **1.000** | 0.640 |
| ollama | 1 | 2 | 1.000 | 1.000 |
| praisy | 29 | 62 | 0.838 | 0.851 |
| supabase | 3 | 9 | 1.000 | 1.000 |
| tasmeen | 2 | 2 | 0.500 | 1.000 |
| vercel | 11 | 6 | 0.795 | 0.727 |

0.015 AUC pooled, and the per-term rows go both ways — MFCC is perfect on `Matthieu` where the encoder is 0.640, and the encoder is perfect on `Tasmeen` where MFCC is chance. Do not read this as the encoder being better. On the data that decides anything, the two are the same.

And on the one mechanism that actually beat its blind control — the two-bank gate on rule substitutions, #92 — the encoder and MFCC score identically at 156 of 194, with 8 of 9 reverts the same substitution.

### What to take from this

1. Do not buy the top rung. A trained speech embedding was the thing the middle rung was supposed to justify. It did not.
2. Do not swap MFCC out. It costs about 1ms per proposal, has no fixed window, no centring step, and needs no CoreML plumbing. Nothing in the app reads encoder states today.
3. The encoder is worth keeping as an instrument. It is much more stable across microphones and sessions, so it is the right tool for asking whether a bank is matching the word or the room. That is how it was used here.
4. The clip bank matches the session more than anyone published. MFCC pooled goes 0.943 → 0.800 when the day is held out. Every AUC in #74, #77, #82 and #87 is a per-clip hold-out number and is therefore optimistic. Say so wherever they are quoted.

### What would falsify it

- Run arm 3 on more terms. Five of the eleven have no correct proposal at all, so they contribute nothing to the 0.810 / 0.825. A set built from labels rather than from proposals would test all eleven.
- A second microphone with a real bank behind it. Arm 1 is two clips. Ten cross-device pairs would turn it into a rate.
- The three single-day terms. `Claude`, `Redcrawl` and `Redrock` cannot enter the leave-one-session-out table at all: #100.

### Where the evidence lives

A scratch directory that no longer exists. Everything above is copied out of it. The scripts were `enc.py` (the CoreML sanity check), `small_test_common.py` (the shared MFCC/DTW/encoder code, with MFCC and DTW copied verbatim from `scripts/reference-matching.py` on `origin/test/abstain-curve`), `small-test.py` / `small-test2.py` / `small-test3.py` (the three runs, the third being the one that ran the encoder over whole utterances), `sweep.py` (the slice sweep), `tables.py` (the 11-row cross-microphone tables), `loso2.py` (the 122×122 distance matrix and the folds) and `control.py` (the size-matched control).

The plan's own entry is `docs/proposals/vocabulary-v3.md`, section "Still open, deliberately not scheduled", on `origin/test/clip-review`.

</details>
